conda_gitenv/deploy.py

```diff
@@ -50,87 +49,61 @@ def deploy_tag(repo, tag_name, target, pkg_cache):
-            lock_name = os.path.join(pkg_cache, dist_name)
-            with Locked(lock_name):
-                schannel_dist_name = dist_name
-                if pkg_info['schannel'] != 'defaults':
-                    schannel_dist_name='{}::{}'.format(pkg_info['schannel'],
-                                                       dist_name)
-                if not conda.install.is_extracted(schannel_dist_name):
-                    if not conda.install.is_fetched(schannel_dist_name):
-                        print('Fetching {}'.format(dist_name))
-                        conda.fetch.fetch_pkg(pkg_info)
-                    conda.install.extract(schannel_dist_name)
-                conda.install.link(target, schannel_dist_name)
+        with open(spec_fname, 'r') as fh:
+            spec = yaml.safe_load(fh)
+
+        channels = prioritize_channels(spec.get('channels', []))
+        # Build reverse look-up from channel URL to channel name.
+        channel_by_url = {url: channel for url, (channel, _) in channels.items()}
+        index = fetch_index(channels, use_cache=False)
+        resolver = Resolve(index)
+        # Create the package distribution from the manifest. Ensure to replace
+        # channel-URLs with channel names, otherwise the fetch-extract may fail.
+        dists = [Dist.from_string(pkg, channel_override=channel_by_url.get(url, url)) for url, pkg in pkgs]
+        # Use the resolver to sort packages into the appropriate dependency
+        # order.
+        sorted_dists = resolver.dependency_sort({dist.name: dist for dist in dists})
+
+        pfe = ProgressiveFetchExtract(index, dists)
+        pfe.execute()
+        mkdir_p(target)
+        txn = UnlinkLinkTransaction.create_from_dists(index, target, (), dists)
+        txn.execute()
 
 
 def deploy_repo(repo, target, desired_env_labels=None):
     # Set pkgs_dirs location to be the specified pkg_cache.
     # Cache settings to be reinstated at the end.
-    import conda
-    orig_package_cache_ = conda.install.package_cache_
+    import conda.base.context
+
     pkg_cache = os.path.join(target, '.pkg_cache')
-    try:
-        # Support conda=4.1.*
-        orig_pkgs_dirs = conda.install.pkgs_dirs
-        conda.install.pkgs_dirs = [pkg_cache]
-    except AttributeError:
-        # Support conda>4.1
-        @property
-        def mocker(self):
-            return [pkg_cache]
-        import conda.base.context
-        orig_pkgs_dirs = conda.base.context.Context.pkgs_dirs
-        # Monkey patch the context instance property.
-        conda.base.context.Context.pkgs_dirs = mocker
-
-    # Empty package cache so that it will reinitialised.
-    conda.install.package_cache_ = {}
+    @property
+    def mocker(self):
+        return (pkg_cache,)
+    orig_pkgs_dirs = conda.base.context.Context.pkgs_dirs
+    # Monkey patch the context pkgs_dirs property to override it
+    # with our custom package cache directory.
+    conda.base.context.Context.pkgs_dirs = mocker
```

Comment:
Wowzers. Is this really the only way to control the package cache!?! I can live with it, but it ain't pretty 😉 💩

Reply:
Totally agree. From what I understand it's now pretty locked down, and this is the only way I can see to subvert the default, which is for `pkgs_dirs` to be the `pkgs` directory in the root environment or the users `~/.conda/pkgs` ... I'd be thankful if you could tell me otherwise.